### ISSUE TITLE
excluding specifc images ?

### ISSUE THREAD

**Issue by benihime91** [bghira/SimpleTuner#507]
Hi, i have prepared a parquet metadata file with `image_path`, `caption`, `height`, `width`. Now my directory of training images contains a lot of images and i only want to train with the images that are present in the metadata parquet file. Is there any way i could exclude the images that are in the parquet metadata .

Using the below config for dataset and the default behaviour seems to pick up all the images that are present in the directory.

```json
[
  {
    "id": "dashshaper--v1.2.5_image_dataset",
    "type": "local",
    "instance_data_dir": "/mnt/data1/ayushman/datasets",
    "caption_strategy": "parquet",
    "metadata_backend": "parquet",
    "parquet": {
      "path": "/mnt/data1/ayushman/projects/dash-diffusion/SimpleTuner/work_dir/dashshaper--v1.2.5_meta_cap_clean.parquet",
      "filename_column": "base_image_path",
      "caption_column": "caption",
      "fallback_caption_column": "tags",
      "width_column": "width",
      "height_column": "height",
      "identifier_includes_extension": true
    },
    "resolution": 1.0,
    "minimum_image_size": 0.75,
    "maximum_image_size": 2.0,
    "target_downsample_size": 1.5,
    "prepend_instance_prompt": false,
    "instance_prompt": null,
    "only_instance_prompt": false,
    "disable": false,
    "cache_dir_vae": "/mnt/data1/ayushman/projects/dash-diffusion/SimpleTuner/work_dir/vae_cache",
    "probability": 1.0,
    "skip_file_discovery": "",
    "preserve_data_backend_cache": false,
    "vae_cache_clear_each_epoch": true,
    "repeats": 1,
    "crop": true,
    "crop_aspect": "preserve",
    "crop_style": "random",
    "resolution_type": "area",
    "text_embeds": "dashshaper--v1.2.5_text_embeds_dataset",
    "cache_dir": "/mnt/data1/ayushman/projects/dash-diffusion/SimpleTuner/work_dir/cache_dir"
  },
  {
    "id": "dashshaper--v1.2.5_text_embeds_dataset",
    "dataset_type": "text_embeds",
    "default": true,
    "type": "local",
    "cache_dir": "/mnt/data1/projects/dash-diffusion/SimpleTuner/work_dir/cache_dir"
  }
]

```

**Comment by bghira** [comment]
not presently, though if you set skip_file_discovery to `vae text metadata aspect` the trainer will not replace your aspect bucket files. you can manually remove the entries you no longer want. this may be tedious unless you automate it

**Comment by sayakpaul** [comment]
Other option would be to handle it at the preprocessing level. You can read the metadata of the files you're reading and then filter them w.r.t some criteria (aspect ratio, resolution, etc.).

**Comment by bghira** [comment]
ok, i've set up a test to reproduce this issue, but i can't. i am seeing these log entries:

```
2024-06-21 10:35:10,149 [DEBUG] (ParquetMetadataBackend) Image /Volumes/ml/datasets/pexels/10908479.png was not found on the backend. Skipping image.
2024-06-21 10:35:10,178 [DEBUG] (ParquetMetadataBackend) Image /Volumes/ml/datasets/pexels/MET-ART_ACM_12_0000.jpg was not found on the backend. Skipping image.
2024-06-21 10:35:10,251 [DEBUG] (ParquetMetadataBackend) Image /Volumes/ml/datasets/pexels/10908479.png was not found on the backend. Skipping image.
2024-06-21 10:35:10,252 [DEBUG] (ParquetMetadataBackend) Image /Volumes/ml/datasets/pexels/MET-ART_ACM_12_0000.jpg was not found on the backend. Skipping image.
```

I copied some images into the dir that do not exist in the parquet table.

then, you can check for the existence of this file inside the aspect caches:

```
% grep 'MET-ART_ACM_12_0000' /Volumes/ml/datasets/pexels/aspect_ratio_bucket_*.json
```

and nothing comes back.

so it seems that the images that weren't in the parquet table are skipped during metadata collection and never added to the aspect indices.

**Comment by bghira** [comment]
```
2024-06-21 10:35:10,220 [INFO] (ParquetMetadataBackend) Image processing statistics: {'total_processed': 79, 'skipped': {'already_exists': 0, 'metadata_missing': 0, 'not_found': 2, 'too_small': 0, 'other': 0}}
```

they are tracked as `not_found` inside this log line. i should update this to `metadata_missing`.

**Comment by benihime91** [comment]
Hi @bghira can you share a sample config. It might be something wrong with my config as well. But i do not see the above logs

**Comment by bghira** [comment]
your configuration looks correct. can you check debug.log

**Comment by benihime91** [comment]
hmm, i started a run and it seems to take all the images present `instance_data_dir` into account while calculation `Generating aspect bucket cache`. I verfied this my looking at the total number of tqdm iterations 😄

**Comment by bghira** [comment]
yes, i think you are correct. it goes through the real filesystem contents to scan for metadata in the parquet.

the expectation is that the dataset will probably have fewer entries than the parquet. i never really considered that someone would have extra files just hanging out in their data pool :D 

you see, a lot of my training is done from S3 buckets directly. if we scan the whole dataset and check if the files exist, it would be an expensive process. but finding which files exist and then discovering their metadata was cheaper. now there are file list caches in simpletuner for all data backends. and this kind of thing is cheaper.

but i'm still not sure i'd want to reverse the flow. it will make the parquet backend then behave very differently from the json backend.

is it extremely problematic to have it the way it currently is? or just visually confusing?

**Comment by benihime91** [comment]
> you see, a lot of my training is done from S3 buckets directly. if we scan the whole dataset and check if the files exist, it would be an expensive process. but finding which files exist and then discovering their metadata was cheaper. now there are file list caches in simpletuner for all data backends. and this kind of thing is cheaper.


hmmm, kinda makes sense. 

But why not just check the metadata files in the parquet. You are directly take the image path from the metadata. If the file does not exists throw an error ?  What i mean is something like the metadata json that kohya trainer uses. It only takes the images whole keys exist in the json file

This is should also make handling large datasets easier ?

**Comment by bghira** [comment]
i don't really understand the suggestion or how it differs from what's currently happening. can you elaborate?

in testing, i often use a dataset of 1M files but i have just a local subset of 10k for convenience. if we scan all 1M metadata rows, this is actually a lot more time to spend checking for files that exist versus going for the 10k that are actually there.

**Comment by bghira** [comment]
also, in `feature/multi-caption` branch, i think the issue of scanning for and encoding missing text embeds is solved.

**Comment by benihime91** [comment]
> in testing, i often use a dataset of 1M files but i have just a local subset of 10k for convenience. if we scan all 1M metadata rows, this is actually a lot more time to spend checking for files that exist versus going for the 10k that are actually there.

yeah this makes sense. but conversely scanning only the metadata would be faster in a use case where there are ~5M and i want to train on only 1M and the metadata has the 1M images I am using (current implementation will scan all the 5M images which would be a slow operation ? ). 
In this way i can actually create different subsets of the whole data and iterate over them faster (don't know if this makes sense, can't think of a better way to explain)

> also, in feature/multi-caption branch, i think the issue of scanning for and encoding missing text embeds is solved.

Umm, can you elaborate ?

**Comment by bghira** [comment]
well, it's not trivial to have both ways in the trainer.

the text embed issue might have been reported by someone else.

please open a new issue if you face further problems
